src/transform/src/analysis/equivalences.rs

```diff
@@ -477,12 +477,148 @@ impl EquivalenceClasses {
+            for expr in class.iter_mut() {
+                self.remap.reduce_expr(expr);
+            }
+            class.sort();
+            class.dedup();
+            // for a class to be interesting we require at least two elements that do not reference the same root.
+            let common_class = class
+                .iter()
+                .map(|x| self.remap.get(x))
+                .reduce(|prev, this| if prev == this { prev } else { None });
+            if class.len() == 1 || common_class != Some(None) {
+                class.clear();
+            }
+        }
+        novel.retain(|c| !c.is_empty());
+        novel
+    }
+
+    /// Derives potentially novel equivalences without regard for minimization.
+    ///
+    /// This is an opportunity to explore equivalences that do not correspond to expression minimization,
+    /// and therefore should not be used in `minimize_once`. They are still potentially important, but
+    /// required additional guardrails to ensure we reach a fixed point.
+    ///
+    /// The implications will be introduced into `self.classes` and will prompt a round of minimization,
+    /// making it somewhat polite to avoid producing outputs that cannot result in novel equivalences.
+    /// For example, before producing a new equivalence, one could check that the involved terms are not
+    /// already present in the same class.
+    fn implications(&self) -> Vec<Vec<MirScalarExpr>> {
+        let mut new_equivalences = Vec::new();
+
+        // If we see `false == IsNull(foo)` we can add the non-null implications of `foo`.
+        let mut non_null = std::collections::BTreeSet::default();
+        for class in self.classes.iter() {
+            if class.iter().any(|c| c.is_literal_false()) {
+                for e in class.iter() {
+                    if let MirScalarExpr::CallUnary {
+                        func: mz_expr::UnaryFunc::IsNull(_),
+                        expr,
+                    } = e
+                    {
+                        expr.non_null_requirements(&mut non_null);
+                    }
+                }
+            }
+        }
+        // If we see `true == foo` we can add the non-null implications of `foo`.
+        // TODO: generalize to arbitrary non-null, non-error literals; at the moment `true == pred` is
+        // an important idiom to identify for how we express predicates.
+        for class in self.classes.iter() {
+            if class.iter().any(|c| c.is_literal_true()) {
+                for expr in class.iter() {
+                    expr.non_null_requirements(&mut non_null);
+                }
+            }
         }
+        // Only keep constraints that are not already known
+        let lit_false = MirScalarExpr::literal_false();
+        let target = self.remap.get(&lit_false);
+        non_null.retain(|c| {
+            let is_null = MirScalarExpr::column(*c).call_is_null();
+            self.remap.get(&is_null) != target
```

Comment:
The logic here seems to check out even if `target == None`, but it's quite surprising. If `target == None` but `is_null` is mapped to something... well, that means `is_null` must be mapped to true, not false... contradicting the `non_null_requirements`, so we should record the contradiction to get reformed when we refresh. But also: weird. And the call to refresh is on L487 is load bearing!

Reply:
We discussed, and concluded that the logic is suboptimal for `None`, in that the constraint should be added when they are both `None`, as it means that the information is certainly not known.

Reply:
Logic updated for this case.